### PR TITLE
android: simplify IPNService lifecycle

### PR DESCRIPTION
**Pull request by oxtoacart**

### CODE REVIEW

File: android/src/main/java/com/tailscale/ipn/MainActivity.kt [hunk 1]
```diff
@@ -267,7 +267,9 @@ class MainActivity : ComponentActivity() {
   override fun onNewIntent(intent: Intent?) {
     super.onNewIntent(intent)
     if (intent?.getBooleanExtra(START_AT_ROOT, false) == true) {
-      navController.popBackStack(route = "main", inclusive = false)
+      if (this::navController.isInitialized) {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/MainActivity.kt:270]: Drive-by, @agottardo noticed a crash here when navController is not initialized. If it's not initialized, that means we're just starting the activity, in which case we'll end up on `main` anyway.

Comment (oxtoacart) [reply android/src/main/java/com/tailscale/ipn/MainActivity.kt]: Moved to its own PR - https://github.com/tailscale/tailscale-android/pull/386

File: android/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -83,16 +83,6 @@
             </intent-filter>
         </activity>
 
-
-        <receiver
```
Comment (oxtoacart) [inline android/src/main/AndroidManifest.xml:87]: I removed this level of indirection. I didn't see why it was needed, and things seem to work without it.

File: android/src/main/java/com/tailscale/ipn/IPNService.kt [hunk 3]
```diff
@@ -27,26 +26,29 @@ open class IPNService : VpnService(), libtailscale.IPNService {
...
-      Libtailscale.requestVPN(this)
-      App.getApplication().setWantRunning(true)
-    }
-    return START_STICKY
-  }
-
-  override public fun close() {
+  override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int =
+      when (intent?.action) {
+        ACTION_STOP_VPN -> {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/IPNService.kt:31]: This allows us to stop the service via an Intent

File: android/src/main/java/com/tailscale/ipn/IPNService.kt [hunk 4]
```diff
@@ -106,7 +108,7 @@ open class IPNService : VpnService(), libtailscale.IPNService {
   }
 
   companion object {
-    const val ACTION_REQUEST_VPN = "com.tailscale.ipn.REQUEST_VPN"
+    const val ACTION_START_VPN = "com.tailscale.ipn.START_VPN"
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/IPNService.kt:111]: Renamed for symmetry/clarity.

File: android/src/main/java/com/tailscale/ipn/QuickToggleService.java [hunk 5]
```diff
@@ -114,13 +114,14 @@ public void onClick() {
     }
 
     private void onTileClick() {
-        boolean act;
+        boolean needsToStop;
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/QuickToggleService.java:117]: Renamed for clarity.

File: android/src/main/java/com/tailscale/ipn/QuickToggleService.java [hunk 6]
```diff
@@ -114,13 +114,14 @@ public void onClick() {
...
         synchronized (lock) {
-            act = getAbleToStartVPN(this) && isRunning;
+            needsToStop = getAbleToStartVPN(this) && isRunning;
+        }
+        if (needsToStop) {
+            App.getApplication().stopVPN();
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/QuickToggleService.java:122]: We can reuse the applications stop and start methods instead of building intents by hand.

File: android/src/main/java/com/tailscale/ipn/ui/viewModel/IpnViewModel.kt [hunk 7]
```diff
@@ -70,17 +68,11 @@ open class IpnViewModel : ViewModel() {
   }
 
   fun startVPN() {
-    val context = App.getApplication().applicationContext
-    val intent = Intent(context, IPNReceiver::class.java)
-    intent.action = IPNReceiver.INTENT_CONNECT_VPN
-    context.sendBroadcast(intent)
+    App.getApplication().startVPN()
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/ui/viewModel/IpnViewModel.kt:71]: We can reuse the applications stop and start methods instead of building intents by hand.

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 8]
```diff
@@ -290,72 +288,26 @@ class App : Application(), libtailscale.AppContext {
...
-          val intent: Intent? = VpnService.prepare(act)
-          if (intent == null) {
-            startVPN()
-          } else {
-            startActivityForResult(act, intent, reqCode)
-          }
-        })
-  }
-
-  fun showURL(act: Activity, url: String?) {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:304]: Drive-by, make linter happy

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 9]
```diff
@@ -290,72 +288,26 @@ class App : Application(), libtailscale.AppContext {
...
-          val builder: CustomTabsIntent.Builder = CustomTabsIntent.Builder()
-          val headerColor = -0xb69b6b
-          builder.setToolbarColor(headerColor)
-          val intent: CustomTabsIntent = builder.build()
-          intent.launchUrl(act, Uri.parse(url))
-        })
-  }
-
-  @get:Throws(Exception::class)
-  val packageCertificate: ByteArray?
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:316]: Drive-by, make linter happy

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 10]
```diff
@@ -290,72 +288,26 @@ class App : Application(), libtailscale.AppContext {
...
+        startVPN()
+      } else {
+        // An intent here means that we need to prompt for permission to be a VPN.
+        startActivityForResult(act, prepareIntent, reqCode)
       }
-      return null
     }
-
-  @Throws(IOException::class)
-  fun openUri(uri: String?, mode: String?): Int? {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:328]: Drive-by, make linter happy

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 11]
```diff
@@ -290,72 +288,26 @@ class App : Application(), libtailscale.AppContext {
...
-    val resolver: ContentResolver = contentResolver
-    return mode?.let { resolver.openFileDescriptor(Uri.parse(uri), it)?.detachFd() }
   }
 
-  fun deleteUri(uri: String?) {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:333]: Drive-by, make linter happy

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 12]
```diff
@@ -290,72 +288,26 @@ class App : Application(), libtailscale.AppContext {
...
-    val resolver: ContentResolver = contentResolver
-    resolver.delete(Uri.parse(uri), null, null)
-  }
-
-  fun notifyFile(uri: String?, msg: String?) {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:338]: Drive-by, make linter happy

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 13]
```diff
@@ -290,72 +288,26 @@ class App : Application(), libtailscale.AppContext {
...
-      viewIntent = Intent(DownloadManager.ACTION_VIEW_DOWNLOADS)
-    }
-    val pending: PendingIntent =
-        PendingIntent.getActivity(this, 0, viewIntent, PendingIntent.FLAG_UPDATE_CURRENT)
-    notify(
-        getString(R.string.file_notification), msg, FILE_CHANNEL_ID, pending, FILE_NOTIFICATION_ID)
-  }
-
-  fun createNotificationChannel(id: String?, name: String?, importance: Int) {
+  private fun createNotificationChannel(id: String?, name: String?, importance: Int) {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:304]: Drive-by, make linter happy

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 14]
```diff
@@ -290,72 +288,26 @@ class App : Application(), libtailscale.AppContext {
...
     val channel = NotificationChannel(id, name, importance)
     val nm: NotificationManagerCompat = NotificationManagerCompat.from(this)
     nm.createNotificationChannel(channel)
   }
 
-  fun notify(
+  private fun notify(
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:310]: Drive-by, make linter happy

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 15]
```diff
@@ -421,12 +373,7 @@ class App : Application(), libtailscale.AppContext {
     return sb.toString()
   }
 
-  fun isTV(): Boolean {
-    val mm = getSystemService(Context.UI_MODE_SERVICE) as UiModeManager
-    return mm.currentModeType == Configuration.UI_MODE_TYPE_TELEVISION
-  }
-
-  fun prepareDownloadsFolder(): File {
+  private fun prepareDownloadsFolder(): File {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:376]: Drive-by, make linter happy

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 16]
```diff
@@ -421,12 +373,7 @@ class App : Application(), libtailscale.AppContext {
     return sb.toString()
   }
 
-  fun isTV(): Boolean {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:424]: Drive-by, make linter happy

File: android/src/main/java/com/tailscale/ipn/IPNService.kt [hunk 17]
```diff
@@ -27,26 +26,34 @@ open class IPNService : VpnService(), libtailscale.IPNService {
...
+          // This should never really happen, but we need when to be
+          // exhaustive.
+          START_NOT_STICKY
+        }
+      }
+
+  override fun close() {
     stopForeground(true)
     Libtailscale.serviceDisconnect(this)
-    App.getApplication().setWantRunning(false)
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/IPNService.kt:49]: Move this to `stopVPN()`. Also, if the Go side stops the VPN, it's already set wantRunning to false.

File: android/src/main/AndroidManifest.xml (hunk 2 above)
Comment (kari-ts) [reply android/src/main/AndroidManifest.xml:87]: Aren't these needed for IPNReceiver to receive these intents from external components (eg Tasker)?

File: android/src/main/java/com/tailscale/ipn/IPNService.kt [hunk 18]
```diff
@@ -27,26 +26,34 @@ open class IPNService : VpnService(), libtailscale.IPNService {
...
-  override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
-    if (intent != null && "android.net.VpnService" == intent.action) {
-      // Start VPN and connect to it due to Always-on VPN
-      val i = Intent(IPNReceiver.INTENT_CONNECT_VPN)
-      i.setPackage(packageName)
-      i.setClass(applicationContext, IPNReceiver::class.java)
-      sendBroadcast(i)
-
-    // If intent is null, the service is restarting because the system is attempting to re-create the killed service. If this is the case, check whether we are able to start the the VPN before starting.
-    } else if (intent != null || getAbleToStartVPN(this.getApplicationContext())) {
```
Comment (kari-ts) [inline android/src/main/java/com/tailscale/ipn/IPNService.kt:39]: don't we need to include this case?

File: android/src/main/AndroidManifest.xml (hunk 2 above)
Comment (oxtoacart) [reply android/src/main/AndroidManifest.xml:87]: Ohh, that's why we had IPNReceiver! Okay, let me add it back in ...

File: android/src/main/java/com/tailscale/ipn/IPNService.kt (hunk 18 above)
Comment (oxtoacart) [reply android/src/main/java/com/tailscale/ipn/IPNService.kt:39]: Yes we do.

**Review by kari-ts — APPROVED** [review]
Looks great. So much more readable.

**Comment by oxtoacart** [comment]
Thanks for the review @kari-ts and @agottardo !
